sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/ApproximatePercentile.scala

```diff
@@ -0,0 +1,302 @@
+
+      // Currently, QuantileSummaries ignores the construction parameter compressThresHold,
+      // which may cause QuantileSummaries to occupy unbounded memory. We have to hack around here
+      // to make sure QuantileSummaries doesn't occupy infinite memory.
+      // TODO: Figure out why QuantileSummaries ignores construction parameter compressThresHold
+      if (summaries.sampled.length >= compressThresHoldBufferLength) compress()
+    }
+
+    /** In-place merges in another PercentileDigest. */
+    def merge(other: PercentileDigest): Unit = {
+      if (!isCompressed) compress()
+      summaries = summaries.merge(other.quantileSummaries)
+    }
+
+    /**
+     * Returns the approximate percentiles of all observation values at the given percentages.
+     * A percentile is a watermark value below which a given percentage of observation values fall.
+     * For example, the following code returns the 25th, median, and 75th percentiles of
+     * all observation values:
+     *
+     * {{{
+     *   val Array(p25, median, p75) = percentileDigest.getPercentiles(Array(0.25, 0.5, 0.75))
+     * }}}
+     */
+    def getPercentiles(percentages: Array[Double]): Array[Double] = {
+      if (!isCompressed) compress()
+      if (summaries.count == 0 || percentages.length == 0) {
+        Array.empty[Double]
+      } else {
+        val result = new Array[Double](percentages.length)
+        var i = 0
+        while (i < percentages.length) {
+          result(i) = summaries.query(percentages(i))
+          i += 1
+        }
+        result
+      }
+    }
+
+    private final def compress(): Unit = {
+      summaries = summaries.compress()
+      isCompressed = true
+    }
+  }
+
+  /**
+   * Serializer  for class [[PercentileDigest]]
+   *
+   * This class is NOT thread safe because usage of ExpressionEncoder is not threadsafe.
+   */
+  class PercentileDigestSerializer {
+
+    // In Scala 2.10, the creation of TypeTag is not thread safe. We need to use ScalaReflectionLock
+    // to protect the creation of this encoder. See SI-6240 for more details.
+    private[this] final val serializer = ScalaReflectionLock.synchronized {
+      ExpressionEncoder[QuantileSummariesData].resolveAndBind()
+    }
+
+    // There are 4 fields in QuantileSummariesData
+    private[this] final val row = new UnsafeRow(4)
+
+    final def serialize(obj: PercentileDigest): Array[Byte] = {
+      val data = new QuantileSummariesData(obj.quantileSummaries)
+      serializer.toRow(data).asInstanceOf[UnsafeRow].getBytes
+    }
+
+    final def deserialize(bytes: Array[Byte]): PercentileDigest = {
+      row.pointTo(bytes, bytes.length)
+      val quantileSummaries = serializer.fromRow(row).toQuantileSummaries
+      new PercentileDigest(quantileSummaries, isCompressed = true)
+    }
+  }
+
+  // An case class to wrap fields of QuantileSummaries, so that we can use the expression encoder
+  // to serialize it.
+  case class QuantileSummariesData(
```

Comment:
QuantileSummaries contains some private var fields, so I think it is better to define a wrapper class instead of changing QuantileSummaries to case class (which is not supposed to contain var fields).